### PR TITLE
Callable: S3 to Sharefile

### PR DESCRIPTION
**Pull request by johncmerfeld**
Adds an `s3_to_sharefile` callable and supporting methods. This could be used to share artifacts of DAG runs with partners.

### Usage
```python
from ea_airflow_util.callables.sharefile import s3_to_sharefile

op = PythonOperator(
    task_id=...,
    dag=...,
    python_callable=s3_to_sharefile,
    op_kwargs={
        "s3_conn_id": "<connection ID>",
        "s3_key": "path/to/file.csv",
        "sf_conn_id": "<connection ID>",
        "sf_folder_path": "/path/to/folder",
    },
)
```

### Other additions
Includes a `disk_to_sharefile` callable that accepts file or directory uplaods

Additionally adds methods to the `SharefileHook` for navigating folders and uploading files.

### Testing
This code was tested on a dev server and successfully copied files from S3 to Sharefile.

### CODE REVIEW

**Review by sleblanc23 — APPROVED** [review]
This looks great to me, thanks @johncmerfeld! Appreciate all the documentation too.

File: ea_airflow_util/providers/sharefile/hooks/sharefile.py [hunk 1]
```diff
@@ -76,6 +78,43 @@ def download(self, item_id, local_path):
...
+        if not self.session:
+            self.get_conn()
+
+        # referencing sharefile's upload logic here
+        # https://api.sharefile.com/samples/python
+        upload_config_uri = f"{self.base_url}/Items({folder_id})/Upload"
+        upload_config_resp = self.session.get(upload_config_uri)
+        upload_config = upload_config_resp.json()
+    
+        upload_uri = ""
```
Comment (jayckaiser) [inline ea_airflow_util/providers/sharefile/hooks/sharefile.py:92]: Are there any cases where `upload_uri` is undefined and it's expected? If not, we can remove line 92.

File: ea_airflow_util/providers/sharefile/hooks/sharefile.py [hunk 2]
```diff
@@ -76,6 +78,43 @@ def download(self, item_id, local_path):
...
+            raise AirflowException(f"Failed to get upload link for {local_file}; Reason: {upload_config['reason']}; Message: {upload_config['message']['value']}")
+
+        files = {Path(local_file).name: open(local_file, 'rb')}
+        resp = self.session.post(upload_uri, files=files)
+        resp.raise_for_status()
+
+
+    def folder_id_from_path(self, folder_path):
+        """Given a complete Sharefile folder path, returns the ID of the leaf folder. If the path is not found, returns `None`"""
+        folders = self.find_folders("allshared")
```
Comment (jayckaiser) [inline ea_airflow_util/providers/sharefile/hooks/sharefile.py:106]: Tell me more about the "allshared" string passed to find_folders.

**Review by jayckaiser — CHANGES_REQUESTED** [review]
I've left some comments that I'd like answered (particularly the "allshared" comment).

File: ea_airflow_util/providers/sharefile/hooks/sharefile.py (hunk 1 above)
Comment (jayckaiser) [reply ea_airflow_util/providers/sharefile/hooks/sharefile.py:92]: Fair enough. I'd argue that NullPointerExceptions are preferable to using an empty string in compiled languages when alerting that something is mal-defined. Could we set upload_uri to None and leave a comment next to it that this condition will never be met?

File: ea_airflow_util/providers/sharefile/hooks/sharefile.py (hunk 2 above)
Comment (jayckaiser) [reply ea_airflow_util/providers/sharefile/hooks/sharefile.py:110]: So I think my point of confusion may relate to a misunderstanding of the ShareFile API. Is "allshared" a built-in keyword, or does it mean something specific to the API or the `find_folders()` method?

Comment (johncmerfeld) [reply ea_airflow_util/providers/sharefile/hooks/sharefile.py:110]: It's a way to tell the Sharefile API to return all resources underneath the root. I tried to explain this with my comment but let me know if I can clarify it

File: ea_airflow_util/providers/sharefile/hooks/sharefile.py (hunk 1 above)
Comment (johncmerfeld) [reply ea_airflow_util/providers/sharefile/hooks/sharefile.py:92]: Happy to set it to `None`

File: ea_airflow_util/providers/sharefile/hooks/sharefile.py (hunk 2 above)
Comment (jayckaiser) [reply ea_airflow_util/providers/sharefile/hooks/sharefile.py:110]: Yeah, I think a description that "allshared" is something Sharefile-specific would be helpful. I didn't realize ShareFile had hard-coded that string as a folder ID. What happens if we create a folder called "allshared"?

Comment (johncmerfeld) [reply ea_airflow_util/providers/sharefile/hooks/sharefile.py:112]: I took another pass at it

**Review by jayckaiser — APPROVED** [review]

**Pull request merged**